sdk/src/main/java/tech/relaycorp/sdk/GatewayClientImpl.kt

```diff
@@ -0,0 +1,211 @@
+            request.replyTo = Messenger(object : Handler(Looper.getMainLooper()) {
+                override fun handleMessage(msg: android.os.Message) {
+                    if (msg.what != REGISTRATION_AUTHORIZATION) {
+                        cont.resumeWithException(Exception("pre-register failed"))
+                        context.unbindService(serviceConnection)
+                        return
+                    }
+                    cont.resume(msg.data.getByteArray("auth")!!)
+                    context.unbindService(serviceConnection)
+                }
+            })
+            Messenger(binder).send(request)
+        }
+    }
+
+    // Messaging
+
+    suspend fun sendMessage(message: OutgoingMessage) {
+        withContext(Dispatchers.IO) {
+            val senderCertificate = getParcelDeliveryAuthorization(
+                message.senderEndpoint,
+                FirstPartyEndpoint.load(message.receiverEndpoint.address)!!
+            )
+
+            val parcel = Parcel(
+                recipientAddress = message.receiverEndpoint.address,
+                payload = message.message,
+                senderCertificate = senderCertificate,
+                messageId = message.id.value,
+                creationDate = message.creationDate,
+                ttl = Duration.between(
+                    message.creationDate,
+                    message.expirationDate
+                ).seconds.toInt(),
+                senderCertificateChain = setOf(
+                    Storage.getCertificate(message.receiverEndpoint.address)!!,
+                    Storage.getGatewayCertificate()!!
+                )
+            )
+
+            val senderPrivateKey = Storage.getKeyPair(message.senderEndpoint.address)!!.private
+            return@withContext try {
+                PoWebClient.initLocal(Relaynet.POWEB_PORT)
+                    .deliverParcel(
+                        parcel.serialize(senderPrivateKey),
+                        Signer(
+                            senderCertificate,
+                            senderPrivateKey
+                        )
+                    )
+                true
+            } catch (e: Exception) {
+                Log.e("SendMessage", "Error sending message", e)
+                false
+            }
+        }
+
+    }
+
+    private val incomingMessageChannel = BroadcastChannel<IncomingMessage>(1)
+    fun receiveMessages(): Flow<IncomingMessage> = incomingMessageChannel.asFlow()
+
+    // Internal
+
+    internal suspend fun handleIncomingMessageNotification() {
+        val wasBound = syncConnection != null
+        if (!wasBound) bind()
+
+        val poweb = PoWebClient.initLocal(Relaynet.POWEB_PORT)
+
+        val nonceSigners = Storage
+            .listEndpoints()
+            .map { endpoint ->
+                Signer(
+                    Storage.getCertificate(endpoint)!!,
+                    Storage.getKeyPair(endpoint)!!.private
+                )
```

Comment:
This would be a good thing to implement as an instance field on `FirstPartyEndpoint` (e.g., `firstPartyEndpoint.getSigner()`), since it'll be used in other places, like parcel delivery.